Lesson2

fimakoff opened the pull request.

fimakoff committed:
Add files frow HW#2

fimakoff committed:
ease refactor.

Saiph41182 commented on lesson2/Main.java:
```diff
@@ -0,0 +1,62 @@
+
+    private static void parseAndSumm(String[][] arr) throws MyArraySizeException, MyArrayDataException {
+        int summ = 0;
+        if (arr.length != 4) throw new MyArraySizeException("Размер массива не соответствует необходимому");
```
Тут есть нюанс. двумерный массив - это массив массивов. Т.е. в каждой из четырех (в данном случае) ячеек лежит еще один массив. И его длина может быть и 10 и 20 и 100500. При том, длина каждого из четырех (в данном случае) может быть разная. К примеру {{1},{1,2},{1,2,3},{1,2,3,4}} если у него вызвать array.length то на выходе будет 4. 
Поэтому надо еще и через цикл проверить каждый вложенный массив, чтобы проверить требование к размеру 4х4. 
Это можно сделать в месте следующего коммента (смотри ниже), чтобы не запускать еще один цикл. Это сэкономит нам n итераций. Где n - количество массивов в двумерном массиве.

Saiph41182 commented on lesson2/Main.java:
```diff
@@ -0,0 +1,62 @@
+    private static void parseAndSumm(String[][] arr) throws MyArraySizeException, MyArrayDataException {
+        int summ = 0;
+        if (arr.length != 4) throw new MyArraySizeException("Размер массива не соответствует необходимому");
+        for (int i = 0; i < arr.length; i++) {
```
Вот тут можно добавить проверку

Saiph41182 commented on lesson2/MyArrayDataException.java:
```diff
@@ -0,0 +1,9 @@
+package lesson2;
+
+class MyArrayDataException extends Exception{
```
Тут можно добавить два поля i,j, чтобы пользователь, у которого возникнет такое исключение, мог его обработать и исправить. Но для этого надо будет добавить еще и геттеры и сеттеры.

Saiph41182 commented on lesson2/Main.java:
```diff
@@ -0,0 +1,62 @@
+        System.out.println("Сумма элементов массива составляет: " + summ);
+    }
+
+    private static boolean isNumeric(String strNum) {
```
Очень находчивое решение))

Saiph41182 commented on lesson2/Main.java:
```diff
@@ -0,0 +1,62 @@
+        };
+        try {
+            parseAndSumm(array1);
+        } catch (MyArraySizeException | MyArrayDataException e) {
```
Такое решение очень конструктивное в случае, если у вас одинаковые требования к обоим исключениям: вывести стектрейс, напечатать сообщение и т.к. Но, если захотеть добавить функционал, как в комментарии к классу ArrayDataException, то тогда придется в операторе catch проверять к какому типу принадлежит exception. А это никому из нас не надо)) Проще будет разделить исключения и добавить еще один оператор catch. 
!!!!Но будьте внимательны. операторы catch зависят от порядка объявления. если первым будет catch(Exception e){ ... } а после него сколь угодно много других исключений наследников exception, то они перехватываться не будут. А будет отлавливаться только первый кэтч. Т.е. Объявлять кэтч стоит в порядке восхождения иерархии: от самого низа к классу exception и throwable.

fimakoff committed:
Добавил проверку для вложенного массива.
Появился затык на реализации метода по изменению элемента массива с некорректным типом данных.

fimakoff committed:
easy ref

fimakoff committed:
Start work!

fimakoff committed:
Добавил возможность корректировки некорректного элемента массива.

fimakoff committed:
Небольшой рефакторинг кода.

Saiph41182 commented on lesson2/MyArrayDataException.java:
```diff
@@ -0,0 +1,13 @@
+        System.out.println(message + i + ", " + j);
+        System.out.print("Введите корретное значение: ");
+        arr[i][j] = sc.next();
+    }
```
Тут, чтобы решить, достаточно ли будет этого функционала, можно задать себе вопрос: " Каким именно способом мы/кто-то будем исправлять не валидные данные в массиве?". В теории мы можем использовать метод parseAndSum где угодно. Может это будет приложение, которое получает данные из интернета, может из файла, а может и с клавиатуры. 
Класс main может быть частью абсолютно любого приложения. Поэтому, будет универсально, если мы выкинем i,j наружу, а сам класс исключения будет их только хранить. Позволяя нам исправить ошибку так, как нам надо. Далее смотри коммент к методу letsgo

Saiph41182 commented on lesson2/Main.java:
```diff
@@ -0,0 +1,72 @@
+        } catch (MyArrayDataException e) {
+            e.getMessage();
+            parseAndSumm(arr);
+        }
```
Сначала лучше причитать коммент снизу. 
Во втором кэтче можно написать что-то вроде: 
arr[e.getI(),e.getJ()] = взять откуда угодно нужное значение;
parseAndSum(arr);

fimakoff committed:
Добавил геттеры на позицию в массиве в MyArrayDataException
Изменил сигнатуру MyArrayDataException
Теперь обработка исключения выполняется во втором catch'е.

fimakoff committed:
Добавил геттеры на позицию в массиве в MyArrayDataException
Изменил сигнатуру MyArrayDataException

fimakoff committed:
Revert "Добавил геттеры на позицию в массиве в MyArrayDataException Изменил сигнатуру MyArrayDataException"

This reverts commit b1fb9192
